🤫 Quiet down railties tests

zzak opened the pull request:
This PR addressed several unrelated issues in the Railties test suite, in order to reduce the amount of noise.

**After**

```
$ bundle exec rake test >> railties_output_patched 2>&1
$ wc -l railties_output_patched
    1491 railties_output_patched
```

**Before**

```
$ bundle exec rake test >> railties_output_unpatched 2>&1
$ wc -l railties_output_unpatched
   19023 railties_output_unpatched
```

zzak committed:
Stub `rails_command` to prevent `app:template` command from being run

When #48269 was merged any gem installed during `rails new` which calls `app:template` would cause the install command to be executed and consequently `bundle install` would also run.

We want to avoid running these commands in our tests because they are very expensive.

It is up to the gem (importmap, etc) to test the behavior of the install command, not railties.

Before

```
$ bin/test test/generators/plugin_generator_test.rb test/generators/app_generator_test.rb
Finished in 320.803659s, 0.8541 runs/s, 7.1913 assertions/s.
274 runs, 2307 assertions, 14 failures, 0 errors, 0 skips
```

After

```
Finished in 70.316250s, 3.9251 runs/s, 34.3164 assertions/s.
276 runs, 2413 assertions, 0 failures, 0 errors, 0 skips
```

zzak committed:
Change railties default log_level to :error

The default of :info generates a ton of unnecessary noise in the railties logs.

skipkayhil commented on railties/test/application/action_controller_test_case_integration_test.rb:
```diff
@@ -33,6 +33,7 @@ def time_zone
     RUBY
 
     remove_from_config '.*config\.load_defaults.*\n'
+    add_to_config "config.active_support.cache_format_version = 7.1"
```
This unfortunately feels like the type of thing we'll inevitably have to update on new releases. Could we template like `Rails::VERSION::STRING.to_f` in here or is that overkill?

zzak replied:
I looked into this, and we can't simply use whatever the current `RAILS_VERSION` is because it may not be supported by AS::Cache:
https://github.com/rails/rails/blob/633eb5a9f4f7ed9e638ac5b6b9ee04f323bb61c0/activesupport/lib/active_support/cache.rb#L759-L763

jonathanhefner replied:
When #49624 is merged, this will be unnecessary (in terms of suppressing warnings), right?  If so, I think it would be better to leave this out.

zzak replied:
👍 Dropped.

skipkayhil commented on railties/test/application/configuration_test.rb:
```diff
@@ -102,6 +103,8 @@ def suppress_sqlite3_warning
 
     def restore_sqlite3_warning
       remove_from_config ".*config.active_record.sqlite3_production_warning.*\n"
+      remove_from_config "config.active_record.sqlite3_production_warning.*"
```
Could these be a single regex instead of two?

skipkayhil commented on railties/test/generators/app_generator_test.rb:
```diff
@@ -677,7 +677,7 @@ def test_web_console
   def test_generation_runs_bundle_install
     generator([destination_root])
-    run_generator_instance
+    run_generator_and_bundler [destination_root]
```
:+1: this one is great

skipkayhil commented on railties/test/generators/shared_generator_tests.rb:
```diff
@@ -344,14 +344,14 @@ def test_dev_option
 
   def test_edge_option
-    Rails.stub(:gem_version, Gem::Version.new("2.1.0")) do
+    Rails.stub(:gem_version, Gem::Version.new("7.1.0")) do
```
This one feels bad as well, it would be great if we didn't have to update it in the future

zzak replied:
Also looking at this one, we could do something like read the `RAILS_VERSION` file, but there are separate tests for alpha and non-alpha, and that file isn't the same on stable branches. So we could craft these strings ourselves, but that seems like overkill for two tiny test cases.

matthewd replied:
The whole point of using an unambiguously ancient version here is that it should never need to change.

zzak replied:
For `test_edge_option` the test generates the following output 9 times, but _only_ when running the `app_generator_test` and not `plugin_generator_test` even though they both include it.

```
warning: rejected HEAD because shallow roots are not allowed to be updated
Could not find gem 'rails' in https://github.com/rails/rails.git (at 2-1-stable@b5d759f).
error: Could not read 52b85b91abbf45a4af37d9383d44bf25a0699a75
```

I will continue to investigate.

matthewd replied:
Yeah -- it doesn't surprise me that actual `2-1-stable` makes bundler unhappy, but this test has no reason to need to be talking to the network / seeing the actual repo contents.

skipkayhil commented on railties/test/generators/shared_generator_tests.rb:
```diff
@@ -411,9 +425,18 @@ def run_generator_using_prerelease(args)
+        # AppBase#run_bundle will add an extra platform for Apple Silicon Macs
+        # So we have to bump the bundle command index
+        # https://github.com/rails/rails/blob/2e4061666ba74cc2b2291152173e7e3a40e28450/railties/lib/rails/generators/app_base.rb#L689-L690
+        if RUBY_PLATFORM.start_with?("arm64")
```
:+1:

matthewd replied:
This conditional feels gross. At absolute minimum we should continue the surrounding pattern and check that the extra command is the one we're expecting... but I'd much rather just not.

Do we just want `@bundle_commands.last`? Or otherwise, can we shift out the commands as we match them, and make this one optional in that process?

zzak replied:
👍 Updated to use shift in a5c7fb86fb185b00d24a83cf3521619ccbd898e1

jonathanhefner replied:
I might be wrong, but I think Matthew's suggestion was to remove the conditional entirely by adding something like 

```ruby 
@bundle_commands.shift while @bundle_commands.first.start_with?("lock --add-platform=")
```

before the final `assert_match`.

Another alternative would be to `join` the commands:

```ruby 
assert_match %r"^exec rails (?:plugin )?new #{Regexp.escape Shellwords.join(expected_args)}", @bundle_commands.join
```

Though that would have less of a guarantee with respect to ordering.

zzak replied:
I thought it's better to keep the assertions, since this is the only place which covers this behavior:
https://github.com/rails/rails/blob/344e59e1d1ca92703aef3947ef1e0e28abb6709c/railties/lib/rails/generators/app_base.rb#L686-L690

Maybe it could be made into a separate test though?

jonathanhefner replied:
@zzak I've fixed this particular issue in #50169 and added you as a co-author.

zzak replied:
Thanks! I will rebase this out then when I get a chance. :bow:

skipkayhil commented on railties/test/isolation/abstract_unit.rb:
```diff
@@ -478,6 +483,7 @@ def use_frameworks(arr)
 
       if to_remove.include?(:activerecord)
         remove_from_config "config.active_record.*"
+        remove_from_config ".*config.active_record.*\n"
```
Same thing as before, any reason to not make these a single line?

zzak replied:
Good call, fixed in 7d74aedee9 and same as above

zzak commented on railties/test/isolation/abstract_unit.rb:
```diff
@@ -255,11 +259,12 @@ def make_basic_app
       @app.config.session_store :cookie_store, key: "_myapp_session"
       @app.config.active_support.deprecation = :log
-      @app.config.log_level = :info
+      @app.config.log_level = :error
```
TFW you realized this was all your fault 9 years ago: #17757

matthewd commented on railties/test/application/rake/multi_dbs_test.rb:
```diff
@@ -786,7 +786,9 @@ class TwoMigration < ActiveRecord::Migration::Current
 
-        cache_size_b = rails("runner", "p AnimalsBase.connection.schema_cache.size", stderr: true).strip
+        output = rails("runner", "p AnimalsBase.connection.schema_cache.size").split("\n")
+        assert_match(/Ignoring .*\.yml because it has expired/, output[0])
```
Is this test really expecting to always see an expired schema cache? Is that an intended/necessary part of the test?

zzak replied:
I think the answers are "Yes, we should expect to see an expired cache in this case", and "No it's not a necessary part of this test".

So I've dropped this commit in favor of #49859 to discuss how to approach that over there.

matthewd commented on railties/test/application/configuration_test.rb:
```diff
@@ -4324,8 +4328,11 @@ def new(app); self; end
+      remove_from_config ".*config.active_support.cache_format_version.*\n"
 
-      app "development"
+      assert_deprecated(ActiveSupport.deprecator) do
```
This feels very broad... probably worth matching the expected deprecation message

zzak replied:
Dropped

jonathanhefner commented on railties/test/application/configuration_test.rb:
```diff
@@ -94,14 +94,18 @@ def suppress_default_config
     end
 
     def suppress_sqlite3_warning
-      add_to_config "config.active_record.sqlite3_production_warning = false"
+      add_to_config <<-RUBY
+        config.active_record.sqlite3_production_warning = false
+      RUBY
```
What is the a reason to change this?

zzak replied:
This was rebased into a single commit in 3e2fc8e2fb7e6c18b75d2a64162273b2a2f3e250, but the goal was to maintain consistent indentation like other examples of `add_to_config`.

For example, when using `add_to_config` without the `<<-`:

```
puts File.read("#{app_path}/config/application.rb")
require_relative "boot"

require "rails/all"

# Require the gems listed in Gemfile, including any gems
# you've limited to :test, :development, or :production.
%w(sprockets/railtie importmap-rails).each { |r| require r }

module AppTemplate
  class Application < Rails::Application
    # Initialize configuration defaults for originally generated Rails version.
    config.load_defaults 7.2

    # Please, add to the `ignore` list any other `lib` subdirectories that do
    # not contain `.rb` files, or that should not be reloaded or eager loaded.
    # Common ones are `templates`, `generators`, or `middleware`, for example.
    config.autoload_lib(ignore: %w(assets tasks))

    # Configuration for the application, engines, and railties goes here.
    #
    # These settings can be overridden in specific environments using the files
    # in config/environments, which are processed later.
    #
    # config.time_zone = "Central Time (US & Canada)"
    # config.eager_load_paths << Rails.root.join("extras")
        config.hosts << proc { true }
        config.eager_load = false
        config.session_store :cookie_store, key: "_myapp_session"
        config.cache_store = :mem_cache_store
        config.active_support.deprecation = :log
        config.action_controller.allow_forgery_protection = false
        config.active_record.sqlite3_production_warning = false
        config.active_support.cache_format_version = 7.1
config.active_record.sqlite3_production_warning = false
config.logger = MyLogRecorder.new




  end
end
```

zzak replied:
This commit can also be dropped if #49715 is merged

zzak replied:
Droppedl

jonathanhefner commented on railties/test/application/url_generation_test.rb:
```diff
@@ -16,6 +16,7 @@ def app
       require "action_view/railtie"
 
       class MyApp < Rails::Application
+        config.active_support.cache_format_version = 7.1
```
Same comment here regarding #49624.

jonathanhefner commented on railties/test/generators/app_generator_test.rb:
```diff
@@ -677,7 +677,7 @@ def test_web_console
 
   def test_generation_runs_bundle_install
     generator([destination_root])
-    run_generator_instance
+    run_generator_and_bundler [destination_root]
```
I'm unclear on the reason for this change.  This test is simply asserting that `bundle install` is invoked.  There doesn't seem to be a need to actually run Bundler.  As I mentioned in https://github.com/rails/rails/pull/47437#issuecomment-1489269862, it would be better to avoid running Bundler when possible for test performance reasons.

Although I should point out that `run_generator_and_bundler` isn't actually running Bundler here.  The call to `generator([destination_root])` on line 679 [sets the `@generator` variable using the given args](https://github.com/rails/rails/blob/f11c6ac45c0bc2a9a3af333a0c6cca97bac62675/railties/lib/rails/generators/testing/behavior.rb#L82).  So the call to `generator` with `"--no-skip-bundle"` inside `run_generator_and_bundler` is ignored.  In other words, __if__ this change is fixing something, it is due to stubbing `rails_gemfile_entry`.  If that is the case, we should extract a helper for that and use that instead.

zzak replied:
Thanks for your review!

I've made [fa86b88](https://github.com/rails/rails/pull/49679/commits/fa86b88d44248bdab1e977d4419dee064a801699) as another approach which stubs things and only tests commands we expect to be called and that's it.

jonathanhefner commented on railties/test/generators/app_generator_test.rb:
```diff
@@ -702,7 +702,7 @@ def test_generation_use_original_bundle_environment
 
   def test_bundler_binstub
     generator([destination_root])
-    run_generator_instance
+    run_generator_and_bundler [destination_root]
```
Same comment here regarding `run_generator_and_bundler`.

jonathanhefner commented on railties/test/generators/app_generator_test.rb:
```diff
@@ -1054,7 +1054,7 @@ def test_after_bundle_callback
     end
 
-    run_generator_instance
+    run_generator_and_bundler [destination_root]
```
Same comment here regarding `run_generator_and_bundler`.  See the call to `generator` on line 1053.  What is this fixing?

jonathanhefner commented on railties/test/generators/shared_generator_tests.rb:
```diff
@@ -148,7 +148,7 @@ def test_template_is_executed_when_supplied_an_https_path
 
     generator.stub(:apply, apply_stub) do
-      run_generator_instance
+      run_generator_and_bundler [destination_root]
```
Same comment here regarding `run_generator_and_bundler`.  See the call to `generator` on line 145.

jonathanhefner commented on railties/test/isolation/abstract_unit.rb:
```diff
@@ -141,7 +141,11 @@ def build_app(options = {})
         config.active_support.deprecation = :log
         config.action_controller.allow_forgery_protection = false
+        config.active_record.sqlite3_production_warning = false
+        config.active_support.cache_format_version = 7.1
```
Same comment here regarding #49624.

jonathanhefner commented on railties/test/isolation/abstract_unit.rb:
```diff
@@ -255,11 +259,12 @@ def make_basic_app
       @app = Class.new(Rails::Application) do
         def self.name; "RailtiesTestApp"; end
       end
+      @app.config.active_support.cache_format_version = 7.1
```
Same comment here regarding #49624.

jonathanhefner commented on railties/test/generators/app_generator_test.rb:
```diff
@@ -728,15 +728,9 @@ def test_skip_active_job_option
+    assert_nil @rails_commands, "`importmap:install` expected to not be called."
 
-    generator.stub(:rails_command, command_check) do
-      run_generator_instance
-    end
+    run_generator_instance
```
Here, you are asserting before running the generator, so `@rails_commands` will always be nil.

zzak replied:
@jonathanhefner Nice catch!

I thought about this, but can you think of an elegant way to assert neither of these commands are called?

```diff
--- a/railties/test/generators/app_generator_test.rb
+++ b/railties/test/generators/app_generator_test.rb
@@ -728,10 +728,11 @@ def test_skip_active_job_option
   def test_skip_javascript_option
     generator([destination_root], skip_javascript: true)

-    assert_nil @rails_commands, "`importmap:install` expected to not be called."
-
     run_generator_instance

+    assert_not_includes @rails_commands, "importmap:install", "`importmap:install` expected to not be called."
+    assert_not_includes @rails_commands, "turbo:install stimulus:install", "`turbo:install stimulus:install` expected to not be called."
```

jonathanhefner commented on railties/test/generators/app_generator_test.rb:
```diff
@@ -863,15 +827,13 @@ def test_skip_javascript_option_with_skip_js_argument
 
   def test_hotwire
     run_generator_and_bundler [destination_root]
+
+    assert_includes @rails_commands, "turbo:install stimulus:install", "`turbo:install stimulus:install` expected to be called, but wasn't."
```
I think I am :+1: on this change because it should be the gem's responsibility to test its installer with various version of Rails, including edge Rails.  But are we sure there is sufficient test coverage in the `turbo-rails` and `stimulus-rails` gems?  Or are they currently relying on the Rails test suite to act as a canary?

Also, if we're not actually running the installers, then running Bundler is no longer necessary, and we should change `run_generator_and_bundler` to `run_generator` for performance.

zzak replied:
> I think I am 👍 on this change because it should be the gem's responsibility to test its installer with various version of Rails, including edge Rails. But are we sure there is sufficient test coverage in the `turbo-rails` and `stimulus-rails` gems? Or are they currently relying on the Rails test suite to act as a canary?

I just took a brief look through `turbo-rails` and `stimulus-rails` repos and saw no sort of test, I did find hotwired/turbo-rails#494 and hotwired/stimulus-rails#125 which added support to `bun` for those scripts but did make no attempt to add such tests, in fact #49241 was the PR which added these types of tests to railties which was sort of the inspiration for this change.

I want to say what we're doing here is _enough_ test coverage, and don't feel motivated to go and add additional test coverage to each of these generators. Open to suggestions here.

> Also, if we're not actually running the installers, then running Bundler is no longer necessary, and we should change `run_generator_and_bundler` to `run_generator` for performance.

It seems the main difference between `run_generator` is calling `generator.invoke_all` (which you added in 945347f97c1eb16a9b690d45c5f3d3739a6e4a47), so I've thought of getting rid of `run_generator_and_bundler` with the following:

```diff
   def test_css_option_with_cssbundling_gem
-    run_generator_and_bundler [destination_root, "--css=postcss"]
+    generator [destination_root], ["--css=postcss"]
+    run_generator_instance
```

zzak replied:
@jonathanhefner Ping, just wanted to see if you had a chance to look at this again, I'm still not sure the best path forward here. But for example, if we really want to keep these tests, we could split them out into an integration test that only runs on main?

jonathanhefner commented on railties/test/generators/app_generator_test.rb:
```diff
@@ -886,10 +848,10 @@ def test_skip_hotwire
 
   def test_css_option_with_asset_pipeline_tailwind
     run_generator_and_bundler [destination_root, "--css=tailwind"]
+
+
+    assert_includes @rails_commands, "tailwindcss:install", "`tailwindcss:install` expected to be called, but wasn't."
```
Same comment here regarding test coverage in the `tailwindcss-rails` gem and changing `run_generator_and_bundler` to `run_generator`.

jonathanhefner replied:
I'm currently working on PRs to add those tests. :smiley:  I started with https://github.com/rails/importmap-rails/pull/206 because `AppGeneratorTest` actually doesn't cover much of the behavior of `importmap:install`.  I also just submitted https://github.com/rails/dartsass-rails/pull/49, but I don't have commit privileges for that repo.

As those PRs are completed I will follow up with Rails PRs to update `AppGeneratorTest`.  And I will be sure to add you as a co-author, @zzak.

@flavorjones Do you think `tailwindcss-rails` would benefit from the approach I took in https://github.com/rails/dartsass-rails/pull/49?  Specifically, the way it tests against freshly generated Rails apps from multiple versions of Rails, including Rails `main` branch?

flavorjones replied:
@jonathanhefner If you mean "using appraisal to matrix test against rails", sure, that would be useful, I think. (There is existing coverage, run against Rails 7.1 and edge, for most of the gem (and a reminder that my involvement in the development of the gem has been minimal, I'm mostly a release manager at this point).)

But I also think the risk of breakage is relatively low, so I don't strongly feel like it needs to be added _right now_.

It seems likely that we'll end up adding some kind of support for engines soon, and I'll want to add matrix coverage at that point for sure. So I'm happy to defer more thorough coverage for a bit.

zzak replied:
I've opened hotwired/turbo-rails#553 and hotwired/stimulus-rails#136 based on Jonathan's work here.

Looking into adding tests to jsbundling-rails and cssbundling-rails, but they currently have no testing so might take a little longer to setup.

jonathanhefner replied:
@zzak Thank you! :smile:  I submitted https://github.com/rails/jsbundling-rails/pull/182 the other day, and I am working on similar tests for `cssbundling-rails`.

zzak replied:
Awesome, thanks for following up!! I will resolve this thread since I think we will have PRs open (or merged) for all of the upstream libraries properly.

jonathanhefner commented on railties/test/generators/app_generator_test.rb:
```diff
@@ -901,8 +863,9 @@ def test_css_option_with_tailwind_uses_cssbundling_gem_when_using_node
 
   def test_css_option_with_asset_pipeline_sass
     run_generator_and_bundler [destination_root, "--css=sass"]
+
+    assert_includes @rails_commands, "dartsass:install", "`dartsass:install` expected to be called, but wasn't."
```
Same comment here regarding test coverage in the `dartsass-rails` gem and changing `run_generator_and_bundler` to `run_generator`.

zzak replied:
Fixed by rails/dartsass-rails#49

jonathanhefner commented on railties/test/generators/app_generator_test.rb:
```diff
@@ -914,8 +877,9 @@ def test_css_option_with_sass_uses_cssbundling_gem_when_using_node
 
   def test_css_option_with_cssbundling_gem
     run_generator_and_bundler [destination_root, "--css=postcss"]
+
+    assert_includes @rails_commands, "css:install:postcss", "`css:install:postcss` expected to be called, but wasn't."
```
Same comment here regarding test coverage in the `cssbundling-rails` gem and changing `run_generator_and_bundler` to `run_generator`.

jonathanhefner commented on railties/test/generators/app_generator_test.rb:
```diff
@@ -925,6 +889,15 @@ def test_css_option_with_cssbundling_gem_does_not_force_jsbundling_gem
     assert_gem "importmap-rails"
   end
 
+  def test_default_generator_executes_all_rails_commands
+    run_generator_and_bundler [destination_root]
+
+    expected_commands = [
+      "credentials:diff --enroll", "importmap:install", "turbo:install stimulus:install"
+    ]
+    assert_equal expected_commands, @rails_commands
+  end
```
I don't think this test is necessary.  [`test_generator_configures_decrypted_diffs_by_default`](https://github.com/rails/rails/blob/121e0ad015c74eaaca82c6c9a85f5b2e0672ca0b/railties/test/generators/app_generator_test.rb#L531-L534) already asserts the result of `credentials:diff --enroll`, and the tests you've modified cover the other commands.  Additionally, this test is brittle.  We would have to update it whenever we add or remove and `rails_command` calls in the app generator.

zzak replied:
The idea was to make sure that inserting a random `rails_command "lol"` somewhere will trip something somewhere in the test suite, and not only when you run it live, same goes for the empty test case.

jonathanhefner commented on railties/test/generators/plugin_generator_test.rb:
```diff
@@ -244,6 +244,13 @@ def test_no_duplicate_gemfile_entries_when_using_prerelease
     end
   end
 
+  def test_generation_runs_rails_commands
+    generator([destination_root])
+    run_generator_instance
+
+    assert_empty @rails_commands
+  end
```
This test also seems unnecessary.  Why would we want to guarantee that the plugin generator doesn't call any `rails` commands?

rafaelfranca committed:
Merge tests that are doing the same thing together

This will make sure we have less tests to run since those are slow.

Generate the app once and run many assertions on it instead of generating
the same app multiple times.

zzak committed:
Remove unnecessary --asset-pipeline=sprockets from app generator test

Since we are stubbing bundle commands, we can avoid sprockets side-effects here.

This commit also fixes the tests for preservation of sprockets during `app:update`.

Due to propshaft being the default, this test was unnecessary.

rafaelfranca commented on railties/test/generators/app_generator_test.rb:
```diff
@@ -244,12 +244,12 @@ def test_app_update_does_not_generate_assets_initializer_when_sprockets_and_prop
 
-  def test_app_update_does_not_generate_manifest_config_when_propshaft_is_used
-    run_generator [destination_root, "-a", "propshaft"]
+  def test_app_update_does_not_generate_manifest_config_when_sprockets_is_used
```
Isn't this test time wrong now? I think this test is making sure the `manifest.js` file isn't deleted right?

zzak replied:
Removing since this test already exists:
https://github.com/zzak/rails/blob/12a625e5cb682b2c180740be6239dec33291b4f1/railties/test/generators/app_generator_test.rb#L331-L340
